### ISSUE TITLE
Platforms bug: bad hosts (mainly logging)

### ISSUE THREAD

**Issue by hjoliver** [cylc/cylc-flow#5719]
The bad_hosts logging seems to be a bit broken.

To reproduce, define a platform with 3 non-existent hosts (and local install target to avoid remote init):

```ini
# global.cylc
[platforms]
    [[fake]]
        hosts = nein, nada, zippo
        job runner = slurm
        install target = localhost
```

Test worklow:
```ini
# flow.cylc
[scheduling]
   [[graph]]
      R1 = "foo"
[runtime]
   [[foo]]
      platform = fake
```

At start-up:
```console
WARNING - platform: None - Could not connect to nein.
    * nein has been added to the list of unreachable hosts
    * jobs-submit will retry if another host is available.
CRITICAL - [1/foo preparing job:01 flows:1] submission failed
INFO - [1/foo preparing job:01 flows:1] => submit-failed
ERROR - Incomplete tasks:
      * 1/foo did not complete required outputs: ['succeeded']
CRITICAL - Workflow stalled
WARNING - PT1H stall timer starts NOW
```

Problems:
- platform name logged is "None"
- it suggests that jobs-submit will retry automatically, but I think that feature is still on the yet-to-be-implemented list?

If I force trigger the task, same result for the next fake host "nada".

Force trigger again, same result for the last host "zippo".

Force trigger again, this time I get:
```console
INFO - [1/foo waiting job:04 flows:1] => preparing
ERROR - [jobs-submit cmd] (remote init)
    [jobs-submit ret_code] 1
CRITICAL - [1/foo preparing job:04 flows:1] submission failed
INFO - [1/foo preparing job:04 flows:1] => submit-failed
CRITICAL - platform: fake - initialisation did not complete (no hosts were reachable)
WARNING - stall timer stopped
ERROR - Incomplete tasks:
      * 1/foo did not complete required outputs: ['succeeded']
CRITICAL - Workflow stalled
WARNING - PT1H stall timer starts NOW
```

Note:
 - platform name now logged correctly
 - but it appears that no job submission occurred because there are no hosts left

Force trigger again: now the bad host list gets cleared (albeit with no log message) and we're back to the start:
```console
...
WARNING - platform: None - Could not connect to zippo.
    * zippo has been added to the list of unreachable hosts
    * jobs-submit will retry if another host is available.
...
```

Problem:
 - we should clear the bad hosts list *before* job submission to avoid the unnecessary submit-fail between running out of hosts and clearing the list.

We probably need a fix for these issues before the platforms refactor detailed in https://github.com/cylc/cylc-flow/issues/5404

The quick fix would not have to sort out automatic retry, just:
- [ ] log the correct platform name
- [ ] log that submission failed because the host was unreachable but you can manually retrigger to try another host
- [ ] clear the bad hosts lists before job submission if there are no good hosts left (not after arbitrarily failing another submission)
- [ ] log when the bad host list is cleared

**Comment by hjoliver** [comment]
(I did confirm that the main loop plugin is correctly clearing the bad_hosts list).

**Comment by wxtim** [comment]
> it suggests that jobs-submit will retry automatically, but I think that feature is still on the yet-to-be-implemented list?

I'm pretty sure I implemented it, so if it ain't working it's a bug!

**Comment by hjoliver** [comment]
Re-running the above test in another environment, it all worked perfectly for me too  🤔. I'll revisit in my work environment tomorrow to see what was going on...

**Comment by oliver-sanders** [comment]
Closed by #5720
